### PR TITLE
Upgrade opentelemetry to 0.18.0

### PR DESCRIPTION
**Pull request by BrynCooke**
Deal with all the changes in APIs from the OtelUpgrade and other miscellaneous improvements.

1. Metrics have changed significantly. Traits have been introduced, so we can now have cleaner integration otel and expose a regular Meter even though we are aggregating under the hood.

2. We can't actually update yet as we need https://github.com/open-telemetry/opentelemetry-rust/issues/894 to land in a release. The ability to send spans to multiple exporters has been removed accidentally from JaegerCollector.

Closes https://github.com/apollographql/router/issues/1948

- [x] Update opentelemetry
- [x] Refactor metrics
- [x] Remove spaceport
- [x] Otel config errors are only logged once.
- [x] Otel traces are not sent if `tracesIgnored` is detected.

### CODE REVIEW

File: apollo-router/src/plugins/telemetry/tracing/jaeger.rs [hunk 1]
```diff
@@ -89,23 +90,34 @@ impl TracingConfigurator for Config {
...
-                .with(username, |b, u| b.with_collector_username(u))
-                .with(password, |b, p| b.with_collector_password(p))
-                .with_collector_endpoint(&endpoint.to_string())
-                .init_async_exporter(opentelemetry::runtime::Tokio)?,
+                // _endpoint,
+                // _username,
+                // _password,
+                ..
+            } =>
+                todo!("waiting for new release of OTel https://github.com/open-telemetry/opentelemetry-rust/issues/894")
```
Comment (bnjjj) [inline apollo-router/src/plugins/telemetry/tracing/jaeger.rs:110]: So we should wait for the next release before merging right ?

**Review by bnjjj — APPROVED** [review]

File: apollo-router/src/plugins/telemetry/tracing/jaeger.rs (hunk 1 above)
Comment (BrynCooke) [reply apollo-router/src/plugins/telemetry/tracing/jaeger.rs:110]: Yep

File: apollo-router/src/plugins/telemetry/otlp.rs [hunk 2]
```diff
@@ -128,26 +127,50 @@ fn option_metadata_map(gen: &mut schemars::gen::SchemaGenerator) -> schemars::sc
...
             .try_with(&config.ca, |b, c| {
                 Ok(b.ca_certificate(tonic::transport::Certificate::from_pem(c)))
             })?
             .try_with(
                 &config.cert.clone().zip(config.key.clone()),
                 |b, (cert, key)| Ok(b.identity(tonic::transport::Identity::from_pem(cert, key))),
-            )
+            );
+
+        tracing::info!("{:?}", tls);
```
Comment (BrynCooke) [inline apollo-router/src/plugins/telemetry/otlp.rs:172]: TODO: Remove

**Comment by abernix** [comment]
Marking this as a draft because I think we're blocked until the release.

**Comment by BrynCooke** [comment]
@bazaglia The issue has been that the latest otel release broke our Jager integration and there hasn't been another release since. I just figured out a workaround though so maybe we can get this over the line.

File: apollo-router/src/plugins/telemetry/apollo.rs [hunk 3]
```diff
@@ -19,19 +21,26 @@ use super::metrics::apollo::studio::SingleStatsReport;
...
 #[derive(Derivative)]
 #[derivative(Debug)]
 #[derive(Clone, Deserialize, JsonSchema)]
 #[serde(deny_unknown_fields)]
 pub(crate) struct Config {
     /// The Apollo Studio endpoint for exporting traces and metrics.
-    #[schemars(with = "Option<String>")]
+    #[schemars(with = "Option<String>", default)]
+    #[serde(default = "endpoint_default")]
     pub(crate) endpoint: Option<Url>,
```
Comment (bnjjj) [inline apollo-router/src/plugins/telemetry/apollo.rs:44]: If it's a default, we should not wrap it into an `Option<>`

File: apollo-router/src/plugins/telemetry/apollo_exporter.rs [hunk 4]
```diff
@@ -140,80 +139,131 @@ impl ApolloExporter {
...
-#[derive(Debug)]
-pub(crate) struct ReporterManager {
-    endpoint: Url,
-}
+        let req = self
+            .client
+            .post(self.endpoint.clone())
+            .body(compressed_content)
+            .header("X-Api-Key", self.apollo_key.clone())
+            .header("Content-Encoding", "gzip")
```
Comment (bnjjj) [inline apollo-router/src/plugins/telemetry/apollo_exporter.rs:188]: Use constant from http crate for header name

File: apollo-router/src/plugins/telemetry/apollo_exporter.rs [hunk 5]
```diff
@@ -140,80 +139,131 @@ impl ApolloExporter {
...
+            .header(CONTENT_TYPE, "application/protobuf")
+            .header("Accept", "application/json")
```
Comment (bnjjj) [inline apollo-router/src/plugins/telemetry/apollo_exporter.rs:190]: Use constant from http crate for header name

File: apollo-router/src/plugins/telemetry/apollo_exporter.rs [hunk 6]
```diff
@@ -140,80 +139,131 @@ impl ApolloExporter {
...
+            .header(
+                "User-Agent",
```
Comment (bnjjj) [inline apollo-router/src/plugins/telemetry/apollo_exporter.rs:192]: Use constant from http crate for header name

File: NEXT_CHANGELOG.md [hunk 7]
```diff
@@ -26,6 +26,61 @@ By [@USERNAME](https://github.com/USERNAME) in https://github.com/apollographql/
 
 # [x.x.x] (unreleased) - 2022-mm-dd
 
+## ❗ BREAKING ❗
+
+### Jaeger scheduled_delay moved to batch_processor->scheduled_delay ([Issue #2232](https://github.com/apollographql/router/issues/2232))
```
[the remark this answers is not in the document]
Comment (BrynCooke) [reply NEXT_CHANGELOG.md:38]: Yep, I added the migration already, but it's still a breaking change.

File: .circleci/config.yml [hunk 8]
```diff
@@ -56,6 +56,10 @@ parameters:
     type: string
     # update this as new versions of jaeger become available
     default: "1.33.0"
+  protoc_version:
+    type: string
+    # update this as new versions of jaeger become available
```
Comment (bazaglia) [inline .circleci/config.yml:61]: Is this comment mentioning Jaeger a leftover from duplicating the `jaeger_version` parameter? This is the protoc needed by the [opentelemetry-proto](https://github.com/open-telemetry/opentelemetry-rust/tree/main/opentelemetry-proto) crate.

File: apollo-router/src/plugins/telemetry/apollo.rs (hunk 3 above)
Comment (BrynCooke) [reply apollo-router/src/plugins/telemetry/apollo.rs]: [3152c3c](https://github.com/apollographql/router/pull/1970/commits/3152c3c206c6f686bfd1ee12ed842c6e26b75af5)

File: apollo-router/src/plugins/telemetry/apollo_exporter.rs (hunk 4 above)
Comment (BrynCooke) [reply apollo-router/src/plugins/telemetry/apollo_exporter.rs]: [6a1155b](https://github.com/apollographql/router/pull/1970/commits/6a1155b6b7fed93e3b20279fa35ce5b3f92a15ee)

File: apollo-router/src/plugins/telemetry/apollo_exporter.rs (hunk 5 above)
Comment (BrynCooke) [reply apollo-router/src/plugins/telemetry/apollo_exporter.rs]: [6a1155b](https://github.com/apollographql/router/pull/1970/commits/6a1155b6b7fed93e3b20279fa35ce5b3f92a15ee)

File: .circleci/config.yml (hunk 8 above)
Comment (BrynCooke) [reply .circleci/config.yml:61]: Yep

Comment (BrynCooke) [reply .circleci/config.yml]: [f44320d](https://github.com/apollographql/router/pull/1970/commits/f44320d31eaa40cb1a91ad46fe31bf541cbaada0)

**Review by bnjjj — APPROVED** [review]

**Pull request merged**